onnxruntime/core/common/threadpool.cc

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "core/platform/threadpool.h"
+
+#ifdef USE_EIGEN_THREADPOOL
+#if defined(_MSC_VER)
+#pragma warning(disable : 4267)
+#endif
+#include <unsupported/Eigen/CXX11/ThreadPool>
+#else
+#include "task_thread_pool.h"
+#endif
+
+namespace onnxruntime {
+
+namespace concurrency {
+#ifdef USE_EIGEN_THREADPOOL
+
+// TODO: This is temporarily taken from Eigen until we upgrade its version.
+// Barrier is an object that allows one or more threads to wait until
+// Notify has been called a specified number of times.
+class Barrier {
+ public:
+  Barrier(unsigned int count) : state_(count << 1), notified_(false) {
+    eigen_assert(((count << 1) >> 1) == count);
+  }
+  ~Barrier() {
+    eigen_assert((state_ >> 1) == 0);
+  }
+
+  void Notify() {
+    unsigned int v = state_.fetch_sub(2, std::memory_order_acq_rel) - 2;
+    if (v != 1) {
+      eigen_assert(((v + 2) & ~1) != 0);
+      return;  // either count has not dropped to 0, or waiter is not waiting
+    }
+    std::unique_lock<std::mutex> l(mu_);
+    eigen_assert(!notified_);
+    notified_ = true;
+    cv_.notify_all();
+  }
+
+  void Wait() {
+    unsigned int v = state_.fetch_or(1, std::memory_order_acq_rel);
+    if ((v >> 1) == 0) return;
+    std::unique_lock<std::mutex> l(mu_);
+    while (!notified_) {
+      cv_.wait(l);
+    }
+  }
+
+ private:
+  std::mutex mu_;
+  std::condition_variable cv_;
+  std::atomic<unsigned int> state_;  // low bit is waiter flag
+  bool notified_;
+};
+
+class ThreadPool::Impl : public Eigen::NonBlockingThreadPool {
+ public:
+  Impl(const std::string& name, int num_threads)
+      : Eigen::NonBlockingThreadPool(num_threads) {}
+
+  void ParallelFor(int32_t total, std::function<void(int32_t)> fn) {
+    // TODO: Eigen supports a more efficient ThreadPoolDevice mechanism
+    // We will simply rely on the work queue and stealing in the short term.
+    Barrier barrier(static_cast<unsigned int>(total));
+    std::function<void(int32_t)> handle_iteration;
+    handle_iteration = [=, &handle_iteration, &barrier, &fn](int iteration) {
+      fn(iteration);
+      barrier.Notify();
+    };
+
+    for (int32_t id = 0; id < total; ++id) {
+      Schedule([=, &handle_iteration]() { handle_iteration(id); });
+    }
+
+    barrier.Wait();
```

Comment:
>barrier.Wait(); [](start = 4, length = 15)

This thread is not doing any useful work or this is the semantics we must obey?

Reply:
If you want more control over scheduling and synchronization, the primitives are there for integration into your kernel.  This method is a simple wrapper that assumes you are willing to wait for the completion of all threads.

Reply:
Typically it will wait after it has done its portion. The way its coded it offloads everything to other threads and itself does not call the fn()

---
In reply to: [268311146](https://github.com/Microsoft/onnxruntime/pull/674#discussion_r268311146) [](ancestors = 268311146)

Reply:
We don't want threads waiting after they have completed their work.  They need to signal, then return to the queue and find new work quickly.  The thread which spawned the work can synchronize on the notifications, or move on to other things if that is possible.  In the default case, they cannot.  They must guard the results of their work until all members have completed.